### PR TITLE
Change: Spam tab to show only relevant spams for the user

### PR DESCRIPTION
**Pull request by cotom972**
Spam fragment displays only SMS spams that exists in the DB and exists also in the user's inbox.

`SpamRecyclerAdapter` holds a list `inboxSpam` intersected from the spams in the DB, compared by body & sender to the user's inbox messages.

Main title / Spam tab title - Displays number of messages found "similar" (body & sender) in the DB.
(Spam tab shows only messages that are in the inbox)
Inbox title - Number of messages found in the inbox.

### CODE REVIEW

File: projects/noloan/app/src/main/java/hasadna/noloan/mainactivity/MainActivity.java [hunk 1]
```diff
@@ -162,17 +164,24 @@ public boolean onNavigationItemSelected(MenuItem item) {
...
+    statusTitle.setText(String.valueOf(SmsMessages.get().countInboxSpam()));
 
-    // Tab's titles
+    // Number of messages in the inbox
     tabLayout
         .getTabAt(0)
         .setText(
             getString(R.string.inboxFragment_title, SmsMessages.get().getInboxMessages().size()));
+
+    // Number of messages that the user had suggested
```
Comment (cotom972) [inline projects/noloan/app/src/main/java/hasadna/noloan/mainactivity/MainActivity.java]: To be fixed next push. This isn't correct. Suppose to be number of messages relevant to the user's inbox.

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java [hunk 2]
```diff
@@ -29,26 +31,80 @@
...
+  ArrayList<SmsMessage> inboxSpam;
+
   public SpamRecyclerAdapter() {
     Handler handler = new Handler(Looper.myLooper());
+    inboxSpam = new ArrayList<>();
+
+    // Fetch spams from DB that are in user's inbox
+    for (int i = 0; i < SmsMessages.get().getInboxMessages().size(); i++) {
+      for (SmsMessage sms : SmsMessages.get().getDbMessages()) {
+        if (SmsMessages.get().getInboxMessages().get(i).getBody().contentEquals(sms.getBody())
```
Comment (gulilo) [inline projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: Why didn't you use 'searchMessage'?

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java [hunk 3]
```diff
@@ -29,26 +31,80 @@
...
+    // Keep the received date of the user's message.
     SmsMessages.get()
         .setMessagesListener(
             new SmsMessages.MessagesListener() {
               @Override
               public void messageAdded(SmsMessage newMessage) {
-                handler.post(() -> notifyItemInserted(SmsMessages.get().getDbMessages().size()));
+                int inboxIndex =
+                    SmsMessages.get()
+                        .searchMessage(newMessage, SmsMessages.get().getInboxMessages());
```
Comment (gulilo) [inline projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: Cache the 'InboxMessages' since it is used in multiple places.
Same for 'DbMessages'.

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java [hunk 4]
```diff
@@ -29,26 +31,80 @@
...
+                                  .getReceivedAt())
+                          .build());
+                  handler.post(() -> notifyItemChanged(inboxSpamIndex));
+                }
               }
 
               @Override
               public void messageRemoved(int index, SmsMessage smsMessage) {
-                handler.post(() -> notifyItemRemoved(index));
+                int inboxSpamIndex = SmsMessages.get().searchMessage(smsMessage, inboxSpam);
```
Comment (gulilo) [inline projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: It can be call index, since there is not other indexes around.
Same in modified.

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java [hunk 5]
```diff
@@ -29,26 +31,80 @@
...
+                }
               }
 
               @Override
               public void messageModified(int index) {
-                handler.post(() -> notifyItemChanged(index));
+                int inboxSpamIndex =
+                    SmsMessages.get()
+                        .searchMessage(SmsMessages.get().getDbMessages().get(index), inboxSpam);
+                if (inboxSpamIndex != -1) {
```
Comment (gulilo) [inline projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: What are you doing here? having hard time following

File: projects/noloan/app/src/main/java/hasadna/noloan/mainactivity/SpamFragment.java [hunk 6]
```diff
@@ -14,13 +14,14 @@
 import noloan.R;
 
 public class SpamFragment extends Fragment {
+  static final String TAG = "SpamFragment";
 
   private OnFragmentInteractionListener fragmentInteractionListener;
   private RecyclerView recyclerView;
-  private SpamRecyclerAdapter spamRecyclerAdapter;
+  SpamRecyclerAdapter spamRecyclerAdapter;
```
Comment (gulilo) [inline projects/noloan/app/src/main/java/hasadna/noloan/mainactivity/SpamFragment.java]: Is it used somewhere?

File: projects/noloan/common/src/main/java/hasadna/noloan/common/SmsMessages.java [hunk 7]
```diff
@@ -42,7 +42,7 @@ public void suggestMessage(SmsMessage smsMessage) {
     // 1. Add user as a "suggester"
     int index = searchDbMessage(smsMessage);
     if ((index != -1)
-        && dbMessages
+        && !dbMessages
```
Comment (gulilo) [inline projects/noloan/common/src/main/java/hasadna/noloan/common/SmsMessages.java]: Why this change?

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java (hunk 2 above)
Comment (cotom972) [reply projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: Good point, forgot about it. Thanks
Fixed :+1:

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java (hunk 3 above)
Comment (cotom972) [reply projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: I got the "hang" of what you were saying about having some `SmsMessages` function statics, and that it's easier to use (instead of `.get()` each time) as I wrote this...

I didn't want to push these changes (concerning SmsMessages static functions) in this PR, so left it like that for now.
Lets raise this idea again in the channel...

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java (hunk 5 above)
Comment (cotom972) [reply projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: It looks a little "messy" to me too, that's why I added the comment above all of these functions. (Add / Modify / Remove)
>// Listen to db smsMessages. Update just relevant messages that are in the inbox as well.
>// Keep the received date of the user's message.

I'll add another comment for this function specifically...

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java (hunk 4 above)
Comment (cotom972) [reply projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: `index` is part of the function's parameter. 

`inboxSpamIndex` feels easier and faster to "grasp" to which list this index belongs to IMO.

File: projects/noloan/app/src/main/java/hasadna/noloan/mainactivity/SpamFragment.java (hunk 6 above)
Comment (cotom972) [reply projects/noloan/app/src/main/java/hasadna/noloan/mainactivity/SpamFragment.java]: MainActivity - `updateTitles()`
Overall I think updating the titles can be done in a better way than it is at the moment, but for a later task...

File: projects/noloan/common/src/main/java/hasadna/noloan/common/SmsMessages.java (hunk 7 above)
Comment (cotom972) [reply projects/noloan/common/src/main/java/hasadna/noloan/common/SmsMessages.java]: It's a bug fix. I think you accidentally removed the `!` here: 
https://github.com/hasadna/monorepo/commit/a14fceba8dc579302d1342bcfb0e83195a269b21#diff-e2258fc75a728aa3ae8880a986424ae0L61

It prevents a user from suggesting a spam more than once.

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java (hunk 4 above)
Comment (gulilo) [reply projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: > index is part of the function's parameter.

Right...
I prefer shorter names so it bother me.

File: projects/noloan/app/src/main/java/hasadna/noloan/mainactivity/SpamFragment.java (hunk 6 above)
Comment (gulilo) [reply projects/noloan/app/src/main/java/hasadna/noloan/mainactivity/SpamFragment.java]: why main activity is the responsible to update the title?
maybe add a update title in each fragment and main activity just call it, so mainActivity don't need to know about the recycler.

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java (hunk 3 above)
Comment (gulilo) [reply projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: getInboxMessages and getDbMessages cannot be static they use the class fields.

just before the call for setMessagesListener (bad name BTW, you are adding not setting) save thous into a variable and use it instead.
calling the same function more then once is not a good thing, It is maybe OK in our case because it is just a get function, but generally it is something you want to avoid.

Plus it make the code longer then it is actually need to be and harder to read.

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java (hunk 5 above)
Comment (gulilo) [reply projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: maybe with the cached list it will be easier to read.
Plus the way it written
as 
a 
single
thing
every
line
kinda make it hard to read. I think ff doing it and it is annoying...

File: projects/noloan/common/src/main/java/hasadna/noloan/common/SmsMessages.java (hunk 7 above)
Comment (gulilo) [reply projects/noloan/common/src/main/java/hasadna/noloan/common/SmsMessages.java]: Opss?

File: projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java [hunk 8]
```diff
@@ -31,24 +31,22 @@
     extends RecyclerView.Adapter<SpamRecyclerAdapter.RecyclerViewHolder> {
   private static final String TAG = "SpamRecyclerAdapter";
 
-  ArrayList<SmsMessage> inboxSpam;
+  List<SmsMessage> inboxSpam;
```
Comment (gulilo) [inline projects/noloan/app/src/main/java/hasadna/noloan/SpamRecyclerAdapter.java]: Why the change from arrayList to List?
